src/string.cr

```diff
@@ -3350,23 +3355,57 @@ class String
     nil
   end
 
+  private macro gen_index_short(int_class, by_char)
+    # simplified Rabin-Karp version with multiplier == 256
+    search_hash = {{int_class}}.new(0)
+    hash = {{int_class}}.new(0)
+    mask = {{int_class}}.new(0)
+
+    search.each_byte do |b|
+      search_hash = (search_hash << 8) | b
+      hash = (hash << 8) | pointer.value
+      mask = (mask << 8) | 0xff
+      pointer += 1
+    end
+    {% if by_char %}
+    search_bytesize = search.bytesize
+    {% end %}
+
+    while true
+      return offset if (hash & mask) == search_hash
+
+      {% if by_char %}
+      char_bytesize = String.char_bytesize_at(pointer - search_bytesize)
+      {% else %}
+      char_bytesize = 1
+      {% end %}
+      return if pointer + char_bytesize > end_pointer
+      case char_bytesize
+      when 1 then update_simplehash 1
+      when 2 then update_simplehash 2
+      when 3 then update_simplehash 3
+      else        update_simplehash 4
+      end
+
+      offset &+= 1
+    end
+  end
```

Comment:
Does this actually need to be a macro?
I'm pretty sure it could be implemented as a def which yields for the current char bytesize (`char_bytesize = yield pointer`). `int_class` can be a regular def parameter.
Additional parameters for `search`, `offset`, `pointer` and `end_pointer` are also needed of course.
The result should be pretty much identical, but the source is a bit mor straightforward and easier to reason about.

Call sites would look like this:
```cr
search_bytesize = search.bytesize
index_short(UInt64) { |pointer| String.char_bytesize_at(pointer - search_bytesize) }
```
```
index_short(UInt32) { 1 }
```

Reply:
I'll try.

Reply:
Done.

It is a bit slower in byte_index.
But since I can't reason why, I think it is ok.
(I tried to look into `objdump -dS`, but code produced by `--release` is purely matched to source)
Let compiler and future contributors improve it.